raft/Raft.tla

```diff
@@ -0,0 +1,247 @@
+    /\ UNCHANGED <<voters, logs, history>>
+
+(* Fetches logs from leader. It's the same as handling MsgAppend.        *)
+FetchLog(v) ==
+    LET leader == softState[v].leader
+    IN /\ leader # None
+       /\ softState[leader].role = "Leader"
+       /\ hardState[leader].term = hardState[v].term
+       /\ LET lastLog == LastLogOf(v)
+          IN IF lastLog \in logs[leader]
+             THEN (* Leader's log matches Follower's. Fetch logs or commit index randomly. *)
+                  \/ \E fetchLastLog \in logs[leader]:
+                        /\ fetchLastLog.index > lastLog.index
+                        /\ logs' = [logs EXCEPT ![v] = @ \union {
+                                      l \in logs[leader]: /\ l.index > lastLog.index
+                                                          /\ l.index <= fetchLastLog.index
+                                   }]
+                        /\ \/ /\ hardState[v].commit # hardState[leader].commit
+                              /\ hardState' = [hardState EXCEPT ![v] = [@ EXCEPT !.commit = Min(fetchLastLog.index, hardState[leader].commit)]]
+                           \/ /\ hardState[v].commit = hardState[leader].commit
+                              /\ UNCHANGED <<hardState>>
+                  \/ /\ hardState[v].commit # lastLog.index
+                     /\ hardState[v].commit < hardState[leader].commit
+                     /\ hardState' = [hardState EXCEPT ![v] = [@ EXCEPT !.commit = Min(hardState[leader].commit, lastLog.index)]]
+                     /\ UNCHANGED <<logs>>
+             ELSE (* Overwrites mismatch logs. *)
+                  /\ lastLog.index > hardState[v].commit
+                  /\ logs' = [logs EXCEPT ![v] = @ \ { lastLog }]
+                  /\ UNCHANGED <<hardState>>
+       /\ UNCHANGED <<voters, softState, history>>
+
+(* Checks if logs of c are up to date. *)
+LogsUpToDate(v, c) == LET lastLog == LastLogOf(v)
+                          candidateLastLog == LastLogOf(c)
+                      IN \/ candidateLastLog.term > lastLog.term
+                         \/ /\ candidateLastLog.term = lastLog.term
+                            /\ candidateLastLog.index >= lastLog.index
+
+(* Votes for any suitable cnadidate. It's the same as handling MsgRequestVote. *)
+Vote(v) ==
+    /\ \E l \in voters: /\ softState[l].role = "Candidate"
+                        /\ \/ hardState[l].term > hardState[v].term
+                           \/ /\ hardState[l].term = hardState[v].term
+                              /\ hardState[v].voteFor = None
+                        /\ LogsUpToDate(v, l)
+                        /\ hardState' = [hardState EXCEPT ![v] = [@ EXCEPT !.voteFor = l, !.term = hardState[l].term]]
+                        /\ softState' = [softState EXCEPT ![v] = [leader |-> None, role |-> "Follower"]]
+    /\ UNCHANGED <<voters, logs, history>>
+
+(* Proposes data as a log entry. *)
+AppendLog(v, data) ==
```

Comment:
Why logs do not use a tuple?

Reply:
What's the difference? To me, structure is more readable.

Reply:
ummm... I just think `Append` in array more matches like `WAL`, but use a set is also ok.

Reply:
Oh, you mean use `Seq` instead of `Set`. I use set to make it easier to store duplicated logs with same index but different term.